src/libs/Navigation/AppNavigator/AuthScreens.js

```diff
@@ -252,7 +252,25 @@ class AuthScreens extends React.Component {
         NetworkConnection.stopListeningForReconnect();
         clearInterval(this.interval);
         this.interval = null;
-        hasLoadedPolicies = false;
+        this.hasLoadedPolicies = false;
+    }
+
+    /**
+     * We want to only load policy info if:
+     * - We are on the free plan beta
+     * - We are not on the transition or new-workspace pages (if we load policies while creating a new one we may accidentally overwrite the new policy due to race condition)
+     * - We have not already loaded policies
```

Comment:
So we just keep calling this method anytime the component updates until all the conditions are met? 

Seems fine but also kind of inside out to me. This might be off base, but I feel like we are starting to get into some territory where Onyx's async philosophy is hurting us in the code readability department.

Would there be any worry about a race condition or even a need to explain it if we could have something like...

```js
const promises = [getBetas()];
if (shouldCreateWorkspace) {
    promises.push(createPolicy());
}

Promise.all(promises)
    .then(() => {
        getPolicyList();
        getPolicySummaries();
    });
```

Not sure if we can actually do this or where this logic would run. We can't do it in `componentDidMount()` because we are probably on `/transition` and not `/workspace/new` when that happens. Then again we don't have to switch to a route called `/workspace/new` and could instead trigger policy creation with a parameter instead.

Reply:
I think I agree, and was suggesting something similar in [this thread](https://expensify.slack.com/archives/C03TQ48KC/p1632373124135400?thread_ts=1632344718.095600&cid=C03TQ48KC). In cases when we're just displaying a full-screen loading indicator while waiting for things to happen, our Onyx async philosophy doesn't seem to serve us well/leads to overly-complicated code.

I think that this whole flow could be simplified by just consolidating everything into a single `/transition` page (i.e: no `workspace/new`). It would just display a loading indicator, and depending on URL params would do any number of things that need to happen before we can actually display the app. It waits for all those things to be done using promises, and then displays the `exitTo` page or falls back on the homepage (optionally with an error growl).

Thoughts?

Reply:
Moved this conversation [here](https://expensify.slack.com/archives/C03TQ48KC/p1632863419353200).